File: distribution/python/dqops/client/models/field_model.py
import datetime
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union, cast

from attrs import define as _attrs_define
from attrs import field as _attrs_field
from dateutil.parser import isoparse

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.parameter_definition_spec import ParameterDefinitionSpec


T = TypeVar("T", bound="FieldModel")


@_attrs_define
class FieldModel:
    """Model of a single field that is used to edit a parameter value for a sensor or a rule. Describes the type of the
    field and the current value.

        Attributes:
            definition (Union[Unset, ParameterDefinitionSpec]): Defines a single field that is a sensor parameter or a rule
                parameter.
            optional (Union[Unset, bool]): Field value is optional and may be null, when false - the field is required and
                must be filled.
            string_value (Union[Unset, str]): Field value for a string field.
            boolean_value (Union[Unset, bool]): Field value for a boolean field.
            integer_value (Union[Unset, int]): Field value for an integer (32-bit) field.
            long_value (Union[Unset, int]): Field value for a long (64-bit) field.
            double_value (Union[Unset, float]): Field value for a double field.
            datetime_value (Union[Unset, datetime.datetime]): Field value for a date time field.
            column_name_value (Union[Unset, str]): Field value for a column name field.
            enum_value (Union[Unset, str]): Field value for an enum (choice) field.
            string_list_value (Union[Unset, List[str]]): Field value for an array (list) of strings.
            integer_list_value (Union[Unset, List[int]]): Field value for an array (list) of integers, using 64 bit
                integers.
            date_value (Union[Unset, datetime.date]): Field value for an date.
    """

    definition: Union[Unset, "ParameterDefinitionSpec"] = UNSET
    optional: Union[Unset, bool] = UNSET
    string_value: Union[Unset, str] = UNSET
    boolean_value: Union[Unset, bool] = UNSET
    integer_value: Union[Unset, int] = UNSET
    long_value: Union[Unset, int] = UNSET
    double_value: Union[Unset, float] = UNSET
    datetime_value: Union[Unset, datetime.datetime] = UNSET
    column_name_value: Union[Unset, str] = UNSET
    enum_value: Union[Unset, str] = UNSET
    string_list_value: Union[Unset, List[str]] = UNSET
    integer_list_value: Union[Unset, List[int]] = UNSET
    date_value: Union[Unset, datetime.date] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        definition: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.definition, Unset):
            definition = self.definition.to_dict()

        optional = self.optional
        string_value = self.string_value
        boolean_value = self.boolean_value
        integer_value = self.integer_value
        long_value = self.long_value
        double_value = self.double_value
        datetime_value: Union[Unset, str] = UNSET
        if not isinstance(self.datetime_value, Unset):
            datetime_value = self.datetime_value.isoformat()

        column_name_value = self.column_name_value
        enum_value = self.enum_value
        string_list_value: Union[Unset, List[str]] = UNSET
        if not isinstance(self.string_list_value, Unset):
            string_list_value = self.string_list_value

        integer_list_value: Union[Unset, List[int]] = UNSET
        if not isinstance(self.integer_list_value, Unset):
            integer_list_value = self.integer_list_value

        date_value: Union[Unset, str] = UNSET
        if not isinstance(self.date_value, Unset):
            date_value = self.date_value.isoformat()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if definition is not UNSET:
            field_dict["definition"] = definition
        if optional is not UNSET:
            field_dict["optional"] = optional
        if string_value is not UNSET:
            field_dict["string_value"] = string_value
        if boolean_value is not UNSET:
            field_dict["boolean_value"] = boolean_value
        if integer_value is not UNSET:
            field_dict["integer_value"] = integer_value
        if long_value is not UNSET:
            field_dict["long_value"] = long_value
        if double_value is not UNSET:
            field_dict["double_value"] = double_value
        if datetime_value is not UNSET:
            field_dict["datetime_value"] = datetime_value
        if column_name_value is not UNSET:
            field_dict["column_name_value"] = column_name_value
        if enum_value is not UNSET:
            field_dict["enum_value"] = enum_value
        if string_list_value is not UNSET:
            field_dict["string_list_value"] = string_list_value
        if integer_list_value is not UNSET:
            field_dict["integer_list_value"] = integer_list_value
        if date_value is not UNSET:
            field_dict["date_value"] = date_value

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.parameter_definition_spec import ParameterDefinitionSpec

        d = src_dict.copy()
        _definition = d.pop("definition", UNSET)
        definition: Union[Unset, ParameterDefinitionSpec]
        if isinstance(_definition, Unset):
            definition = UNSET
        else:
            definition = ParameterDefinitionSpec.from_dict(_definition)

        optional = d.pop("optional", UNSET)

        string_value = d.pop("string_value", UNSET)

        boolean_value = d.pop("boolean_value", UNSET)

        integer_value = d.pop("integer_value", UNSET)

        long_value = d.pop("long_value", UNSET)

        double_value = d.pop("double_value", UNSET)

        _datetime_value = d.pop("datetime_value", UNSET)
        datetime_value: Union[Unset, datetime.datetime]
        if isinstance(_datetime_value, Unset):
            datetime_value = UNSET
        else:
            datetime_value = isoparse(_datetime_value)

        column_name_value = d.pop("column_name_value", UNSET)

        enum_value = d.pop("enum_value", UNSET)

        string_list_value = cast(List[str], d.pop("string_list_value", UNSET))

        integer_list_value = cast(List[int], d.pop("integer_list_value", UNSET))

        _date_value = d.pop("date_value", UNSET)
        date_value: Union[Unset, datetime.date]
        if isinstance(_date_value, Unset):
            date_value = UNSET
        else:
            date_value = isoparse(_date_value).date()

        field_model = cls(
            definition=definition,
            optional=optional,
            string_value=string_value,
            boolean_value=boolean_value,
            integer_value=integer_value,
            long_value=long_value,
            double_value=double_value,
            datetime_value=datetime_value,
            column_name_value=column_name_value,
            enum_value=enum_value,
            string_list_value=string_list_value,
            integer_list_value=integer_list_value,
            date_value=date_value,
        )

        field_model.additional_properties = d
        return field_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/services/check/mapping/models/FieldModel.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.services.check.mapping.models;

import com.dqops.metadata.fields.ParameterDefinitionSpec;
import com.dqops.utils.exceptions.DqoRuntimeException;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Model of a single field that is used to edit a parameter value for a sensor or a rule. Describes the type of the field and the current value.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "FieldModel", description = "Model of a single field that is used to edit a parameter value for a sensor or a rule. Describes the type of the field and the current value.")
public class FieldModel implements Cloneable {
    @JsonPropertyDescription("Field name that matches the field name (snake_case) used in the YAML specification.")
    private ParameterDefinitionSpec definition;

    @JsonPropertyDescription("Field value is optional and may be null, when false - the field is required and must be filled.")
    private boolean optional;

    @JsonPropertyDescription("Field value for a string field.")
    private String stringValue;

    @JsonPropertyDescription("Field value for a boolean field.")
    private Boolean booleanValue;

    @JsonPropertyDescription("Field value for an integer (32-bit) field.")
    private Integer integerValue;

    @JsonPropertyDescription("Field value for a long (64-bit) field.")
    private Long longValue;

    @JsonPropertyDescription("Field value for a double field.")
    private Double doubleValue;

    @JsonPropertyDescription("Field value for a date time field.")
    private LocalDateTime datetimeValue;

    @JsonPropertyDescription("Field value for a column name field.")
    private String columnNameValue;

    @JsonPropertyDescription("Field value for an enum (choice) field.")
    private String enumValue;

    @JsonPropertyDescription("Field value for an array (list) of strings.")
    private List<String> stringListValue;

    @JsonPropertyDescription("Field value for an array (list) of integers, using 64 bit integers.")
    private List<Long> integerListValue;

    @JsonPropertyDescription("Field value for an date.")
    private LocalDate dateValue;

    /**
     * Creates a selective deep/shallow clone of the object. Definition objects are not cloned, but all other editable objects are.
     * @return Cloned instance.
     */
    public FieldModel cloneForUpdate() {
        try {
            FieldModel cloned = (FieldModel) super.clone();
            if (cloned.definition != null) {
                cloned.definition = cloned.definition.deepClone();
            }
            if (cloned.stringListValue != null) {
                cloned.stringListValue = new ArrayList<>(cloned.stringListValue);
            }
            if (cloned.integerListValue != null) {
                cloned.integerListValue = new ArrayList<>(cloned.integerListValue);
            }

            return cloned;
        }
        catch (CloneNotSupportedException ex) {
            throw new DqoRuntimeException("Clone not supported: " + ex.toString(), ex);
        }
    }

    /**
     * Sets a sample value on the correct field value. Copies and converts sample values from the list of sample values.
     * The method is meant to be called when generating the YAML samples for documentation.
     */
    public void applySampleValues() {
        if (this.definition == null || this.definition.getSampleValues() == null || this.definition.getSampleValues().size() == 0) {
            return;
        }

        List<String> sampleValues = this.definition.getSampleValues();

        switch (this.definition.getDataType()) {
            case string_type:
                this.stringValue = sampleValues.get(0);
                return;

            case boolean_type:
                this.booleanValue = Boolean.parseBoolean(sampleValues.get(0));
                return;

            case integer_type:
                this.integerValue = Integer.valueOf(sampleValues.get(0));
                return;

            case long_type:
                this.longValue = Long.valueOf(sampleValues.get(0));
                return;

            case double_type:
                this.doubleValue = Double.valueOf(sampleValues.get(0));
                return;

            case datetime_type:
                this.datetimeValue =LocalDateTime.parse(sampleValues.get(0));
                return;

            case date_type:
                this.dateValue = LocalDate.parse(sampleValues.get(0));
                return;

            case column_name_type:
                this.columnNameValue = sampleValues.get(0);
                return;

            case enum_type:
                this.enumValue = sampleValues.get(0);
                return;

            case string_list_type:
                this.stringListValue = new ArrayList<>(sampleValues);
                return;

            case integer_list_type:
                this.integerListValue = sampleValues.stream().map(sample -> Long.valueOf(sample)).collect(Collectors.toList());
                return;
        }
    }

    /**
     * Returns the field value for the parameter type in the field definition.
     * @return Field value.
     */
    @JsonIgnore
    public Object getValue() {
        switch (this.definition.getDataType()) {
            case string_type:
                return this.stringValue;

            case boolean_type:
                return this.booleanValue;

            case integer_type:
                return this.integerValue;

            case long_type:
                return this.longValue;

            case double_type:
                return this.doubleValue;

            case datetime_type:
                return this.datetimeValue;

            case date_type:
                return this.dateValue;

            case column_name_type:
                return this.columnNameValue;

            case enum_type:
                return this.enumValue;

            case string_list_type:
                return this.stringListValue;

            case integer_list_type:
                return this.integerListValue;

            default:
                throw new DqoRuntimeException("Cannot retrieve a value for the type " + this.definition.getDataType());
        }
    }

    /**
     * Sets an object as a field value (in the model) for a matching data type.
     * @param value Value to store.
     */
    @JsonIgnore
    public void setValue(Object value) {
        switch (this.definition.getDataType()) {
            case string_type:
                this.stringValue = value == null || value instanceof String ? (String)value : value.toString();
                return;

            case boolean_type:
                this.booleanValue = value == null || value instanceof Boolean ? (Boolean)value : Boolean.parseBoolean(value.toString());
                return;

            case integer_type:
                this.integerValue = value == null || value instanceof Integer ? (Integer)value : Integer.valueOf(value.toString());
                return;

            case long_type:
                this.longValue = value == null || value instanceof Long ? (Long)value : Long.valueOf(value.toString());
                return;

            case double_type:
                this.doubleValue = value == null || value instanceof Double ? (Double)value : Double.valueOf(value.toString());
                return;

            case datetime_type:
                this.datetimeValue = value == null || value instanceof LocalDateTime ? (LocalDateTime)value : LocalDateTime.parse(value.toString());
                return;

            case date_type:
                this.dateValue = value == null || value instanceof LocalDate ? (LocalDate)value : LocalDate.parse(value.toString());
                return;

            case column_name_type:
                this.columnNameValue = value == null || value instanceof String ? (String)value : value.toString();
                return;

            case enum_type:
                this.enumValue = value == null || value instanceof String ? (String)value : value.toString();
                return;

            case string_list_type:
                this.stringListValue =  value == null || value instanceof ArrayList<?> ? (ArrayList)value : List.of(value.toString().split(","));
                return;

            case integer_list_type:
                this.integerListValue = value == null || value instanceof ArrayList<?> ? (ArrayList) value :
                        List.of(value.toString().split(",")).stream().map(val -> Long.valueOf(val)).collect(Collectors.toList());
                return;

            default:
                throw new DqoRuntimeException("Cannot retrieve a value for the type " + this.definition.getDataType());
        }
    }
}
